In [71]:
import warnings
warnings.simplefilter('ignore')

import pandas as pd
from google.colab import drive
drive.mount('/content/drive')
data = pd.read_csv('/content/drive/MyDrive/magister_Python/2023_june_raw.csv', sep=";") #Попереднє згрупування колонок у одну відбулося на етапі ексель файлу функцією CONCAT

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [72]:
data.head()

Timestamp         Ваша основна зайнятість в ІТ зараз...  \
0  5.26.2023 13:40:35  Працюю full-time в ІТ-компанії чи ІТ-відділі   
1  5.26.2023 13:40:36  Працюю full-time в ІТ-компанії чи ІТ-відділі   
2  5.26.2023 13:41:02  Працюю full-time в ІТ-компанії чи ІТ-відділі   
3  5.26.2023 13:41:19  Працюю full-time в ІТ-компанії чи ІТ-відділі   
4  5.26.2023 13:41:34  Працюю full-time в ІТ-компанії чи ІТ-відділі   

   Зарплата / дохід у $$$ за місяць лише ставка після сплати податків  \
0                                             2300.0                    
1                                             3450.0                    
2                                              900.0                    
3                                             4000.0                    
4                                             2700.0                    

                        Спеціалізація Ваш тайтл на цій посаді  \
0         Marketing PR SEO Copywriter                  Senior   
1                   Software Engineer                  Middle   
2         Marketing PR SEO Copywriter               Team Lead   
3                   Software Engineer                  Middle   
4  Analyst (Business Data System etc)                  Middle   

  Вкажіть вашу спеціалізацію Game Design Вкажіть вашу спеціалізацію Sound  \
0                                    NaN                              NaN   
1                                    NaN                              NaN   
2                                    NaN                              NaN   
3                                    NaN                              NaN   
4                                    NaN                              NaN   

  Вкажіть вашу спеціалізацію - QA Оберіть вашу посаду Designer / Artist  \
0                             NaN                                   NaN   
1                             NaN                                   NaN   
2                             NaN                                   NaN   
3                             NaN                                   NaN   
4                             NaN                                   NaN   

  Якщо ви працюєте в GameDev  оберіть вашу спеціалізацію  ...  \
0                                                NaN      ...   
1                                                NaN      ...   
2                                                NaN      ...   
3                                                NaN      ...   
4                                                NaN      ...   

  В якій сфері проєкт в якому ви зараз працюєте?  \
0                                           Інша   
1                           Medtech / Healthcare   
2                                     E-commerce   
3                                     E-commerce   
4                           Medtech / Healthcare   

  Основний напрям роботи компанії в якій працюєте  \
0                                   Аутсорсингова   
1                                   Аутсорсингова   
2                                      Продуктова   
3                                   Аутстафінгова   
4                                      Продуктова   

  Кількість спеціалістів у вашій компанії (в Україні/з України)  \
0                                             до 200              
1                                         понад 1000              
2                                             до 200              
3                                         понад 1000              
4                                            до 1000              

  Загальний стаж роботи за нинішньою ІТ-спеціальністю  \
0                                                4.0    
1                                                4.0    
2                                                5.0    
3                                                5.0    
4                                                3.0    

                                   Яка у вас освіта? Знання англійської 

Прибираємо рядки, де сукупна кількість NaN значень більша за 50% від спостережень

In [73]:
threshold = len(data) * 0.5

# Drop columns with more NaN values than the threshold
data_cleaned = data.dropna(axis=1, thresh=threshold)
# Drop column Threshold as it can not give any value
data_cleaned = data_cleaned.drop("Timestamp", axis=1)
data_cleaned.head()

Ваша основна зайнятість в ІТ зараз...  \
0  Працюю full-time в ІТ-компанії чи ІТ-відділі   
1  Працюю full-time в ІТ-компанії чи ІТ-відділі   
2  Працюю full-time в ІТ-компанії чи ІТ-відділі   
3  Працюю full-time в ІТ-компанії чи ІТ-відділі   
4  Працюю full-time в ІТ-компанії чи ІТ-відділі   

   Зарплата / дохід у $$$ за місяць лише ставка після сплати податків  \
0                                             2300.0                    
1                                             3450.0                    
2                                              900.0                    
3                                             4000.0                    
4                                             2700.0                    

                        Спеціалізація Ваш тайтл на цій посаді  \
0         Marketing PR SEO Copywriter                  Senior   
1                   Software Engineer                  Middle   
2         Marketing PR SEO Copywriter               Team Lead   
3                   Software Engineer                  Middle   
4  Analyst (Business Data System etc)                  Middle   

  Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?  \
0                                                NaN                           
1                                   Так використовую                           
2                                                NaN                           
3                                   Так використовую                           
4                                   Так використовую                           

  Основна мова програмування Вкажіть вашу основну спеціалізацію  \
0                        NaN                                NaN   
1                       Dart                   Mobile  розробка   
2                        NaN                                NaN   
3                 JavaScript                Front-end  розробка   
4                        SQL       Робота з даними аналіз даних   

  В якій сфері проєкт в якому ви зараз працюєте?  \
0                                           Інша   
1                           Medtech / Healthcare   
2                                     E-commerce   
3                                     E-commerce   
4                           Medtech / Healthcare   

  Основний напрям роботи компанії в якій працюєте  \
0                                   Аутсорсингова   
1                                   Аутсорсингова   
2                                      Продуктова   
3                                   Аутстафінгова   
4                                      Продуктова   

  Кількість спеціалістів у вашій компанії (в Україні/з України)  \
0                                             до 200              
1                                         понад 1000              
2                                             до 200              
3                                         понад 1000              
4                                            до 1000              

   Загальний стаж роботи за нинішньою ІТ-спеціальністю  \
0                                                4.0     
1                                                4.0     
2                                                5.0     
3                                                5.0     
4                                                3.0     

                                   Яка у вас освіта? Знання англійської мови  \
0  Вища (бакалавр спеціаліст магістр) - одна або ...                Advanced   
1  Вища (бакалавр спеціаліст магістр) - одна або ...        Pre-Intermediate   
2  Вища (бакалавр спеціаліст магістр) - одна або ...      Upper-Intermediate   
3  Вища (бакалавр спеціаліст магістр) - одна або ...      Upper-Intermediate   
4  Вища (бакалавр спеціаліст магістр) - одна або ...      Upper-Intermediate   

  Де ви зараз живете?  В якій області ви зараз живете? Ваша стать  Ваш вік  
0            В Україні                 Київ чи 

Подивимось на описову характеристику категоріальних змінних

In [74]:
categorical_features = data_cleaned.select_dtypes(include=['object'])
categorical_features.describe()

Ваша основна зайнятість в ІТ зараз...      Спеціалізація  \
count                                          14059              13985   
unique                                             9                 27   
top     Працюю full-time в ІТ-компанії чи ІТ-відділі  Software Engineer   
freq                                           12928               6605   

       Ваш тайтл на цій посаді  \
count                    13985   
unique                      10   
top                     Middle   
freq                      5015   

       Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?  \
count                                               11724                           
unique                                                  4                           
top                                      Так використовую                           
freq                                                 8684                           

       Основна мова програмування Вкажіть вашу основну спеціалізацію  \
count                        9172                               9184   
unique                         35                                 10   
top                    JavaScript                 Back-end  розробка   
freq                         1680                               2775   

       В якій сфері проєкт в якому ви зараз працюєте?  \
count                                           13950   
unique                                           1396   
top                                              Інша   
freq                                             1488   

       Основний напрям роботи компанії в якій працюєте  \
count                                            13451   
unique                                               7   
top                                      Аутсорсингова   
freq                                              5446   

       Кількість спеціалістів у вашій компанії (в Україні/з України)  \
count                                               13451              
unique                                                  6              
top                                                до 200              
freq                                                 3702              

                                        Яка у вас освіта?  \
count                                               13388   
unique                                                  9   
top     Вища (бакалавр спеціаліст магістр) - одна або ...   
freq                                                 8622   

       Знання англійської мови Де ви зараз живете?   \
count                    10432                10506   
unique                       6                    3   
top         Upper-Intermediate            В Україні   
freq                      4440                 9556   

       В якій області ви зараз живете? Ваша стать  
count                             9482      10393  
unique                              24          2  
top                    Київ чи область    Чоловік  
freq                              3781       8082

Заповнимо змінні, що мають NA значення, випадковими данними з існуючого датасету (іншим варіантом є заповнити їх найбільш частішими данними(модоюю), для числових змінних, ще можливий варіант - медіана

Варіант використання підстановок випадкових(з існуючих вже) значень

In [95]:
df = pd.DataFrame(data_cleaned)
import random

def fill_nan_randomly(column):
    non_nan_values = column.dropna().unique()
    num_nan = column.isnull().sum()
    random_values = [random.choice(non_nan_values) for _ in range(num_nan)]
    column[column.isnull()] = random_values

# Apply the function to each transformed column
transformed_columns = df.select_dtypes(include=['object','float']).columns
for column in transformed_columns:
    fill_nan_randomly(df[column])

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14059 entries, 0 to 14058
Data columns (total 17 columns):
 #   Column                                                                      Non-Null Count  Dtype  
---  ------                                                                      --------------  -----  
 0   Ваша основна зайнятість в ІТ зараз...                                       14059 non-null  object 
 1   Зарплата / дохід у $$$ за місяць лише ставка після сплати податків          14059 non-null  float64
 2   Спеціалізація                                                               14059 non-null  object 
 3   Ваш тайтл на цій посаді                                                     14059 non-null  object 
 4   Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?  14059 non-null  object 
 5   Основна мова програмування                                                  14059 non-null  object 
 6   Вкажіть вашу основну спеціалізацію            

Варіант використання моди

In [96]:
# Function to fill NaN values with the mode value of the column
def fill_nan_with_mode(column):
    mode_value = column.mode().values[0]  # Get the mode value of the column
    column.fillna(mode_value, inplace=True)  # Fill NaN values with the mode value

# Apply the function to each transformed column
transformed_columns = df.select_dtypes(include=['object', 'float']).columns
for column in transformed_columns:
    fill_nan_with_mode(df[column])


In [91]:
df.head()

Ваша основна зайнятість в ІТ зараз...  \
0  Працюю full-time в ІТ-компанії чи ІТ-відділі   
1  Працюю full-time в ІТ-компанії чи ІТ-відділі   
2  Працюю full-time в ІТ-компанії чи ІТ-відділі   
3  Працюю full-time в ІТ-компанії чи ІТ-відділі   
4  Працюю full-time в ІТ-компанії чи ІТ-відділі   

   Зарплата / дохід у $$$ за місяць лише ставка після сплати податків  \
0                                             2300.0                    
1                                             3450.0                    
2                                              900.0                    
3                                             4000.0                    
4                                             2700.0                    

                        Спеціалізація Ваш тайтл на цій посаді  \
0         Marketing PR SEO Copywriter                  Senior   
1                   Software Engineer                  Middle   
2         Marketing PR SEO Copywriter               Team Lead   
3                   Software Engineer                  Middle   
4  Analyst (Business Data System etc)                  Middle   

  Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?  \
0                                                Так                           
1                                   Так використовую                           
2                                 Ні не використовую                           
3                                   Так використовую                           
4                                   Так використовую                           

  Основна мова програмування Вкажіть вашу основну спеціалізацію  \
0              Pascal/Delphi                    Робота з даними   
1                       Dart                   Mobile  розробка   
2         Flex / Flash / AIR                Full Stack розробка   
3                 JavaScript                Front-end  розробка   
4                        SQL       Робота з даними аналіз даних   

  В якій сфері проєкт в якому ви зараз працюєте?  \
0                                           Інша   
1                           Medtech / Healthcare   
2                                     E-commerce   
3                                     E-commerce   
4                           Medtech / Healthcare   

  Основний напрям роботи компанії в якій працюєте  \
0                                   Аутсорсингова   
1                                   Аутсорсингова   
2                                      Продуктова   
3                                   Аутстафінгова   
4                                      Продуктова   

  Кількість спеціалістів у вашій компанії (в Україні/з України)  \
0                                             до 200              
1                                         понад 1000              
2                                             до 200              
3                                         понад 1000              
4                                            до 1000              

   Загальний стаж роботи за нинішньою ІТ-спеціальністю  \
0                                                4.0     
1                                                4.0     
2                                                5.0     
3                                                5.0     
4                                                3.0     

                                   Яка у вас освіта? Знання англійської мови  \
0  Вища (бакалавр спеціаліст магістр) - одна або ...                Advanced   
1  Вища (бакалавр спеціаліст магістр) - одна або ...        Pre-Intermediate   
2  Вища (бакалавр спеціаліст магістр) - одна або ...      Upper-Intermediate   
3  Вища (бакалавр спеціаліст магістр) - одна або ...      Upper-Intermediate   
4  Вища (бакалавр спеціаліст магістр) - одна або ...      Upper-Intermediate   

  Де ви зараз живете?  В якій області ви зараз живете? Ваша стать  Ваш вік  
0            В Україні                 Київ чи 

In [97]:
transformed_categorical_columns = ['Ваша основна зайнятість в ІТ зараз...',
       'Спеціалізація', 'Ваш тайтл на цій посаді',
       'Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?',
       'Основна мова програмування', 'Вкажіть вашу основну спеціалізацію',
       'В якій сфері проєкт в якому ви зараз працюєте?',
       'Основний напрям роботи компанії в якій працюєте',
       'Кількість спеціалістів у вашій компанії (в Україні/з України)',
       'Яка у вас освіта?', 'Знання англійської мови', 'Де ви зараз живете? ',
       'В якій області ви зараз живете?', 'Ваша стать']

Трансформуємо категоріальні змінні за допомогою One-hot encoding. Побудуємо модель випадкового лісу для визначення найбільш значущих змінних

In [98]:
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split

# Assuming 'transformed_categorical_columns' contains the names of your transformed categorical columns
encoder = OneHotEncoder(sparse=False, drop='first')
X_encoded = encoder.fit_transform(df[transformed_categorical_columns])
encoded_columns = encoder.get_feature_names_out(input_features=transformed_categorical_columns)
X_encoded = pd.DataFrame(X_encoded, columns=encoded_columns)

# Combine encoded features with other numeric features
X = pd.concat([X_encoded, df.drop(transformed_categorical_columns, axis=1)], axis=1)

# Exclude the target feature from the features used for training
y = df['Зарплата / дохід у $$$ за місяць лише ставка після сплати податків']
X = X.drop('Зарплата / дохід у $$$ за місяць лише ставка після сплати податків', axis=1)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Fit a RandomForestRegressor model
model = RandomForestRegressor()
model.fit(X_train, y_train)

# Get feature importances
feature_importances = model.feature_importances_

# Create a DataFrame to display feature importances
importance_df = pd.DataFrame({'Feature': X.columns, 'Importance': feature_importances})

# Sort by importance (descending order)
importance_df = importance_df.sort_values(by='Importance', ascending=False)

print("Feature Importances:")
print(importance_df)


Feature Importances:
                                                Feature  Importance
1511  В якій області ви зараз живете?_Вінниця чи обл...    0.219821
1056  В якій сфері проєкт в якому ви зараз працюєте?...    0.082107
1534  Загальний стаж роботи за нинішньою ІТ-спеціаль...    0.075187
1535                                            Ваш вік    0.055093
74                Основна мова програмування_TypeScript    0.054299
...                                                 ...         ...
498   В якій сфері проєкт в якому ви зараз працюєте?...    0.000000
265   В якій сфері проєкт в якому ви зараз працюєте?...    0.000000
1383  В якій сфері проєкт в якому ви зараз працюєте?...    0.000000
1161  В якій сфері проєкт в якому ви зараз працюєте?...    0.000000
237   В якій сфері проєкт в якому ви зараз працюєте?...    0.000000

[1536 rows x 2 columns]


Приведемо категоріальні змінні, трансформовані раніше, до сумарних значень задля визначення значущості конкретних, заданих спочатку змінних

In [99]:
import pandas as pd

# Initialize a dictionary to store summed importance scores for each original feature
summed_importance = {}

# Iterate through the transformed categorical columns
for transformed_column in transformed_categorical_columns:
    # Find all one-hot encoded columns generated from the original feature
    relevant_columns = [col for col in importance_df['Feature'] if col.startswith(transformed_column)]

    # Sum the importance scores of these relevant columns
    total_importance = importance_df[importance_df['Feature'].isin(relevant_columns)]['Importance'].sum()

    # Store the summed importance score for the original feature
    summed_importance[transformed_column] = total_importance

# Manually add the "Ваш вік" column as a numeric feature
summed_importance['Ваш вік'] = importance_df.loc[importance_df['Feature'] == 'Ваш вік', 'Importance'].values[0]

# Create a DataFrame from the summed importance dictionary
summed_importance_df = pd.DataFrame(list(summed_importance.items()), columns=['Feature', 'Summed Importance'])

# Sort the DataFrame by importance scores in descending order
summed_importance_df = summed_importance_df.sort_values(by='Summed Importance', ascending=False)

# Display the DataFrame
print("Summed Feature Importances:")
print(summed_importance_df)


Summed Feature Importances:
                                              Feature  Summed Importance
12                    В якій області ви зараз живете?           0.273178
6      В якій сфері проєкт в якому ви зараз працюєте?           0.138914
4                          Основна мова програмування           0.130270
7     Основний напрям роботи компанії в якій працюєте           0.078147
2                             Ваш тайтл на цій посаді           0.060919
14                                            Ваш вік           0.055093
5                  Вкажіть вашу основну спеціалізацію           0.051779
8   Кількість спеціалістів у вашій компанії (в Укр...           0.042444
1                                       Спеціалізація           0.027922
9                                   Яка у вас освіта?           0.018479
13                                         Ваша стать           0.015306
10                            Знання англійської мови           0.011739
0               Ваша ос

Найбільш суттєвими змінними (в обох випадках заповнення NA значень) стали:

1) В якій області ви зараз живете?           0.27317

2) В якій сфері проєкт в якому ви зараз працюєте?           0.138914

3) Основна мова програмування           0.130270

Інші змінні мали різні місця в рейтингу важливості при повторюваннях прогонки моделі.

Таким чином, найбільше впливають на заробітну плату айтівців такі параметри, як місцезнаходження, сфера поточного проекту та основна мова програмування